scripts/parse_cd.py

```diff
@@ -0,0 +1,211 @@
+    'Data truncated for column .* at row (\d+)')
+
+cn_tpl = "{{Citation needed|date=%s}}" % pd.datetime.now().strftime("%B %Y")
+
+#logger = logging.getLogger('parse_live')
+#setup_logger_to_stderr(logger)
+
+def section_name_to_anchor(section):
+    # See Sanitizer::escapeId
+    # https://doc.wikimedia.org/mediawiki-core/master/php/html/classSanitizer.html#ae091dfff62f13c9c1e0d2e503b0cab49
+    section = section.replace(' ', '_')
+    # urllib.quote interacts really weirdly with unicode in Python2:
+    # https://bugs.python.org/issue23885
+    section = urllib.parse.quote(e(section), safe = e(''))
+    section = section.replace('%3A', ':')
+    section = section.replace('%', '.')
+    return section
+
+def query_revids(revids):
+    session = mwapi.Session('https://en.wikipedia.org/', 'citationdetective', formatversion=2)
+
+    response_docs = session.post(action='query',
+                                prop='revisions',
+                                rvprop='ids|content',
+                                revids='|'.join(map(str, revids)),
+                                format='json',
+                                utf8='',
+                                rvslots='*',
+                                continuation=True)
+
+    for doc in response_docs:
+        for page in doc['query']['pages']:
+            pageid = page['pageid']
+            revid = page['revisions'][0]['revid']
+            if 'title' not in page:
+                continue
+            title = d(page['title'])
+            text = page['revisions'][0]['slots']['main']['content']
+            if not text:
+                continue
+            text = d(text)
+            yield (revid, pageid, title, text)
+
+self = types.SimpleNamespace() # Per-process state
+
+def initializer(backdir):
+    self.backdir = backdir
+
+    self.wiki = yamwapi.MediaWikiAPI(WIKIPEDIA_API_URL, cfg.user_agent)
+    self.parser = snippet_parser.create_snippet_parser(self.wiki, cfg)
+    self.exception_count = 0
+    """
+    if cfg.profile:
+        self.profiler = cProfile.Profile()
+        self.profiler.enable()
+        # Undocumented :( https://stackoverflow.com/questions/24717468
+        multiprocessing.util.Finalize(None, finalizer, exitpriority=16)
+    """
+def finalizer():
+    self.profiler.disable()
+    profile_path = os.path.join(self.backdir, 'profile-%s' % os.getpid())
+    pstats.Stats(self.profiler).dump_stats(profile_path)
+    stats_path = os.path.join(self.backdir, 'stats-%s' % os.getpid())
+    with open(stats_path, 'wb') as stats_f:
+        pickle.dump(self.parser.stats, stats_f)
+
+def process(df):
+    rows = []
+    results = query_revids(df.rev_id.unique())
+    for revid, pageid, title, wikitext in results:
+        for sentence in df[df.rev_id == revid].statement.values:
+            start = wikitext.find(sentence)
+            if start >= 0:
+                end = start + len(sentence)
+                wikitext = wikitext[:end] + cn_tpl + wikitext[end:]
```

Comment:
Here's an idea that I'm sorry I didn't think of before. The way CH's snippet_parser works is that it:

1- Finds the {{ cn }} template in the text you give it.
2- Inserts a [special HTML marker](https://github.com/eggpi/citationhunt/blob/master/snippet_parser/core.py#L169) around the template.
3- Sends the resulting text to the MediaWiki API to convert it all to HTML.
4- Looks for the marker in the result to figure out where the {{ cn }} used to be.
5- Grabs some text/a paragraph around that point to form a snippet.

Maybe rather than inserting a fake {{ cn }} here, which causes the problem of accidentally having two {{ cn }} (and ultimately lies to the user, since there may not be a {{ cn }} when they go look on Wikipedia), we should just do something similar.

That is, we change the snippet_parser to put a marker around the sentence, which then gets treated like a {{ cn }}: we build a snippet around whatever comes back from the API with a marker. The resulting snippet may or may not have a {{ cn }}, but that's fine.

What do you think?

Reply:
I didn't know the MediaWIki API can convert wikitext to HTML, cool! So, for autodetected sentences, we start from step 2 in the workflow to inserts a HTML maker around the sentence and then process the article as usual, right? That sounds good and it can prevent multiple {{cn}}.

How if the sentence has a \<ref\> already? Does snippet parser just skip it?

Reply:
Yup, that's what I mean. It might turn out to be complicated but it's worth a shot :)

Currently, we don't care about `<ref>` at all in CH. The snippet parser just extracts a snippet around citation needed templates, and sometimes the resulting snippets may have references too ([example](https://tools.wmflabs.org/citationhunt/en?id=3d5022c9)).

But we could totally just skip sentences that have a `<ref>` tag in the CD input I guess. No strong opinion yet, I think we could start by being naive and just importing the CD sentences as-is, and see what the result looks like before we decide to do more filtering?

Reply:
Ok! I'll just modify a little to do something like this:
```
start = wikitext.find(sentence)
marked = _CITATION_NEEDED_MARKER_MARKUP.format(sent = sentence)
if start >= 0:
    end = start + len(sentence)
    wikitext = wikitext[:start] + marked + wikitext[end:]
```